### PR TITLE
Convert bash script to Python 

### PR DESCRIPTION
**Pull request by texasaggie97**
### Summary of Changes

* Uses python3 and included packages only
* Designed to be easy to add additional configurations
* Configure complete
* Verify complete
* Similar functionality to the bash script
* Add `--dry-run` parameter - will not make any changes to the system (both installing/removing packages and updating config files)

### Justification

* [AB#2831708](https://dev.azure.com/ni/94b22d7b-ad7b-4f5e-88f0-867910f91c94/_workitems/edit/2831708)
* [AB#2769020](https://dev.azure.com/ni/94b22d7b-ad7b-4f5e-88f0-867910f91c94/_workitems/edit/2769020)

### Testing

Manual run with nilrt

### Procedure

* [ ] ~~This PR: changes user-visible behavior, fixes a bug, or impacts the project's security profile; and so it includes a [CHANGELOG](https://github.com/ni/nilrt-snac/blob/master/docs/CHANGELOG.md) note.~~
* [X] I certify that the contents of this pull request complies with the [Developer Certificate of Origin](https://github.com/ni/nilrt-snac/blob/master/docs/CONTRIBUTING.md).

### CODE REVIEW

**Comment by texasaggie97** [comment]
If the verify commands are correct, this may allow closing AB#2769020

File: nilrt_snac/_Configs/_ConfigFile.py
Comment (amstewart) [inline nilrt_snac/_Configs/_ConfigFile.py:1]: It's a little confusing that this utility class is in the `_Configs` module when it is not a config. Can it be moved to into the main package?

File: nilrt_snac/_Configs/_BaseConfig.py
Comment (amstewart) [inline nilrt_snac/_Configs/_BaseConfig.py:1]: I think modules (including individual files like this) are normally named in the python `lower_case` styling, rather than like the classes they contain.

File: nilrt_snac/_Configs/_CryptsetupConfig.py
Comment (amstewart) [inline nilrt_snac/_Configs/_CryptsetupConfig.py:9]: `_CryptSetupConfig`?

Also
* `_NIAuthConfig`
* `_NTPConfig`
* `_OPKGConfig`
* `_USBGuardConfig`
* `_WIFIConfig`

File: nilrt_snac/_Configs/_ConsoleConfig.py
Comment (amstewart) [inline nilrt_snac/_Configs/_ConsoleConfig.py]: Why not just `nirtcfg --get ...` here? I don't think `verify` should modify the system state.

File: nilrt_snac/_Configs/_CryptsetupConfig.py
Comment (amstewart) [inline nilrt_snac/_Configs/_CryptsetupConfig.py:16]: No dry run?

File: nilrt_snac/_Configs/_NiauthConfig.py
Comment (amstewart) [inline nilrt_snac/_Configs/_NiauthConfig.py:13]: Why is this touched in the NIAuthConfig?

File: nilrt_snac/_Configs/_NiauthConfig.py
Comment (amstewart) [inline nilrt_snac/_Configs/_NiauthConfig.py]: No dry run?

File: nilrt_snac/_Configs/_NtpConfig.py
Comment (amstewart) [inline nilrt_snac/_Configs/_NtpConfig.py:27]: I was thinking that the dry-run's purpose was to aid devloper testing on non-NILRT systems (hence why it skips the prereq steps). But these `configure()` methods seem to keep modifying the system state in breaking ways. What is the purpose of dry-run then?

File: nilrt_snac/_Configs/_NtpConfig.py
Comment (amstewart) [inline nilrt_snac/_Configs/_NtpConfig.py:13]: A note: by loading the config files in the Config.`__init__()`s, it means that any two configuration classes that modify the same config file will only end up writing the latest class' changes. Because they will all load the original (pre-run) version of the config file. You might move this logic into the `configure()` and `verify()` methods instead, and `save()` at the and of each method execution.

File: nilrt_snac/_Configs/_OpkgConfig.py
Comment (amstewart) [inline nilrt_snac/_Configs/_OpkgConfig.py]: Leftover implementation?

**Review by amstewart — CHANGES_REQUESTED** [review]

File: .gitignore
Comment (texasaggie97) [reply .gitignore:7]: Removed .venv

File: Makefile
Comment (texasaggie97) [reply Makefile:82]: What is the benefit of that? It seems like overkill.

File: nilrt_snac/__main__.py
Comment (texasaggie97) [reply nilrt_snac/__main__.py]: Changed

File: nilrt_snac/__main__.py
Comment (texasaggie97) [reply nilrt_snac/__main__.py]: Consent has been moved to `_configure()`. `_configure()` and `_verify()` changed to return `Errors.<error>.value`

File: nilrt_snac/__init__.py
Comment (texasaggie97) [reply nilrt_snac/__init__.py:17]: Fixed

File: nilrt_snac/_common.py
Comment (texasaggie97) [reply nilrt_snac/_common.py]: Because it is called *check*_iptables. Is it considered a configuration?

File: nilrt_snac/_common.py
Comment (texasaggie97) [reply nilrt_snac/_common.py]: I disagree strongly. From [PEP-8](https://peps.python.org/pep-0008/) - "Use 4 spaces per indentation level."

File: nilrt_snac/__init__.py
Comment (texasaggie97) [reply nilrt_snac/__init__.py:21]: According to PEP-8:
* Should use Error - https://peps.python.org/pep-0008/#exception-names
* Should use CapWords, aka PascalCase - https://peps.python.org/pep-0008/#class-names

File: nilrt_snac/_Configs/__init__.py
Comment (texasaggie97) [reply nilrt_snac/_Configs/__init__.py:1]: Yes, removed

File: nilrt_snac/_verify.py
Comment (texasaggie97) [reply nilrt_snac/_verify.py:16]: Yes, changed

File: nilrt_snac/_Configs/__init__.py
Comment (texasaggie97) [reply nilrt_snac/_Configs/__init__.py:17]: Yes, should be all caps: https://peps.python.org/pep-0008/#constants
Changed

File: nilrt_snac/_configure.py
Comment (texasaggie97) [reply nilrt_snac/_configure.py:8]: Moved

File: nilrt_snac/_verify.py
Comment (texasaggie97) [reply nilrt_snac/_verify.py:8]: Moved

File: nilrt_snac/_Configs/_BaseConfig.py
Comment (texasaggie97) [reply nilrt_snac/_Configs/_BaseConfig.py:5]: Since we do not care about python2 compatibility, added.

File: nilrt_snac/_Configs/_ConfigFile.py
Comment (texasaggie97) [reply nilrt_snac/_Configs/_ConfigFile.py:1]: I like to scope classes and objects at the lowest required level. Only Configs use this class which is why I have it in the _Configs folder. I can move it if you have a strong preference.

File: nilrt_snac/_Configs/_ConsoleConfig.py
Comment (texasaggie97) [reply nilrt_snac/_Configs/_ConsoleConfig.py]: Yes, that was an error and is fixed.

File: nilrt_snac/_Configs/_NiauthConfig.py
Comment (texasaggie97) [reply nilrt_snac/_Configs/_NiauthConfig.py:13]: Removed

File: nilrt_snac/_Configs/_CryptsetupConfig.py
Comment (texasaggie97) [reply nilrt_snac/_Configs/_CryptsetupConfig.py:16]: OpkgHelper knows what the dry_run flag's value is and only install/removed when dry_run is False.

File: nilrt_snac/_Configs/_NiauthConfig.py
Comment (texasaggie97) [reply nilrt_snac/_Configs/_NiauthConfig.py]: See other comment about OpkgHelper.

File: nilrt_snac/_Configs/_NtpConfig.py
Comment (texasaggie97) [reply nilrt_snac/_Configs/_NtpConfig.py:27]: ConfigFile.save() takes a dry_run parameter. Until save is called, the changes are only in memory and are only written to disk if dry_run is False when save is called.

File: nilrt_snac/_Configs/_OpkgConfig.py
Comment (texasaggie97) [reply nilrt_snac/_Configs/_OpkgConfig.py]: Already removed.

File: Makefile
Comment (amstewart) [reply Makefile:82]: You would be able to remove eschew the `find` actions and rely on python to choose what files it thinks are appropriate to include.

File: nilrt_snac/_common.py
Comment (amstewart) [reply nilrt_snac/_common.py]: I think so. It would expect it to be a part of the _NFTablesConfig.

File: nilrt_snac/__init__.py
Comment (amstewart) [reply nilrt_snac/__init__.py:21]: > According to PEP-8:
> 
>     * Should use Error - https://peps.python.org/pep-0008/#exception-names

"... (if the exception actually is an error)."

Is this exception always an error?

>     * Should use CapWords, aka PascalCase - https://peps.python.org/pep-0008/#class-names

I don't think they mean that you cannot capitalize initializations. In fact, in the "Type Variable Names" section immediately after the one you linked, they proscribe CapWords to type variables, and their example includes `VT_contra`.

Also consider the common linux standard library `http.server` (all lower case) and its primary class [HTTPServer](https://docs.python.org/3/library/http.server.html#http.server.HTTPServer).

File: nilrt_snac/_Configs/_BaseConfig.py
Comment (texasaggie97) [reply nilrt_snac/_Configs/_BaseConfig.py:1]: All are renamed.

File: nilrt_snac/_Configs/_CryptsetupConfig.py
Comment (texasaggie97) [reply nilrt_snac/_Configs/_CryptsetupConfig.py:9]: All are renamed

File: nilrt_snac/_common.py
Comment (amstewart) [reply nilrt_snac/_common.py]: That's only a proscription for python code. I'm talking about the printed strings.

File: nilrt_snac/_Configs/_ConfigFile.py
Comment (texasaggie97) [reply nilrt_snac/_Configs/_ConfigFile.py:12]: `path` can be a `pathlib.Path` or a `str`. Updated type hint to show both options.

File: nilrt_snac/_verify.py
Comment (texasaggie97) [reply nilrt_snac/_verify.py:10]: Is a mix of prints and logger output acceptable? If it is, then I am fine with changing this.

File: nilrt_snac/__main__.py
Comment (texasaggie97) [reply nilrt_snac/__main__.py:50]: I think the answer to this question is related to the question below about changing `logger.info()` to `print()`. If the answer there is yes, change to print, then changing here seems reasonable.

File: nilrt_snac/_Configs/_NtpConfig.py
Comment (amstewart) [reply nilrt_snac/_Configs/_NtpConfig.py:27]: OK. I guess that makes sense. I think I saw other places in the Config files where package installation or removal is performed regardless of dry_run. I understand that mimicing that behavior might mean that you cannot dry run subsequent steps that use those package contents; but I think it is more important that the dry-run not make changes to your system state.

File: nilrt_snac/_Configs/_ConfigFile.py
Comment (amstewart) [reply nilrt_snac/_Configs/_ConfigFile.py]: Nah. I suppose that's fine. Dropped.

File: Makefile
Comment (texasaggie97) [reply Makefile:82]: How would we do this without needing to install additional packages?

File: nilrt_snac/__init__.py
Comment (texasaggie97) [reply nilrt_snac/__init__.py:21]: > Is this exception always an error?

It returns one of the values in the `Errors` enum, so I would say yes, it is an error.

Elsewhere in PEP-8 says "Note: When using acronyms in CapWords, capitalize all the letters of the acronym. Thus HTTPServerError is better than HttpServerError."

Given this I think the name should be `SNACError`.

File: nilrt_snac/_verify.py
Comment (texasaggie97) [reply nilrt_snac/_verify.py]: Also, I used the same style as the bash script. I.e. `log INFO "Message"` became `logger.info("Message")`

File: nilrt_snac/_common.py
Comment (texasaggie97) [reply nilrt_snac/_common.py]: Those weren't meant to still be there. I quit indenting like that and thought I had fixed up them all. I missed the ones in the pre_reqs. Removed.

File: nilrt_snac/_common.py
Comment (texasaggie97) [reply nilrt_snac/_common.py]: The issue I see with making it a config is that it yes, it installs a package (iptables) but then checks that the module loaded correctly. This is different than all the other configs. 

None of the other configs can return one of the `Errors.EX*` errors. In this respect it is more like a check than a config, though on the other hand, none of the other checks will change the state of the system.

Why nftables_config? It is changing iptables or are they the same thing?

Comment (amstewart) [reply nilrt_snac/_common.py]: For 24.5, we would only change iptables. But in 25.0, we'll use nftables and firewalld in that task to set things up. So we'll either need to merge this check and that Config now or later.

File: Makefile
Comment (amstewart) [reply Makefile:82]: I think it's just the standardlib `build` module; right?

File: nilrt_snac/_verify.py
Comment (amstewart) [reply nilrt_snac/_verify.py]: Yeah; mixing stdout and stderr is fine. Though you should know (because it's a linux secret) that they are not guaranteed to be well-ordered between each other. So you might have a few lines of stdout print before some stderr lines that were really emitted between each other.

File: nilrt_snac/_common.py
Comment (texasaggie97) [reply nilrt_snac/_common.py]: I am going to leave it as a check for now since that is how the bash script treated it. Once `nftables` is implemented, this can merge in with that.

File: nilrt_snac/__init__.py
Comment (texasaggie97) [reply nilrt_snac/__init__.py]: Changed to `SNACError`.

File: nilrt_snac/_verify.py
Comment (texasaggie97) [reply nilrt_snac/_verify.py]: Everything that was `INFO` is now a print.

File: Makefile
Comment (texasaggie97) [reply Makefile]: This can be a future improvement.

**Comment by texasaggie97** [comment]
# Patch V2

* All review comments addressed either by fixing or explaining
* Follow naming conventions consistently
  * Files names changed to lower case and snake_case
  * `SnacError` -> `SNACError`
  * `Configs` -> `CONFIGS`
* Moves `_configure()` and `_verify()` to `__main__.py` instead of in their own files
* `logger.info()` changed to print
* "verbosity" simplified

File: nilrt_snac/__main__.py
Comment (dmondrik) [inline nilrt_snac/__main__.py:86]: I don't see an `add_parser` for quiet. Is this no longer relevant?

File: nilrt_snac/__main__.py
Comment (dmondrik) [inline nilrt_snac/__main__.py:68]: I don't think it would make sense to specify both `--configure` and `--verify`. Should you check if they have both, and error out?

File: src/nilrt-snac
Comment (dmondrik) [inline src/nilrt-snac:3]: This may be a shell-ism but I don't understand. It seems getting the path and invoking the script are (should be) different things on different lines.

Comment (amstewart) [reply src/nilrt-snac:3]: Variables can be defined for a specific command in bash by preambling them to the command on the same line. eg.
```bash
foo=bar env | grep foo  # prints "foo=bar"
```

It's logically the same as doing
```bash
export PYTHONPATH=${PYTHONPATH}:/usr/local/lib/nilrt-snac
python3 -m nilrt_snac $*
export -n PYTHONPATH
```

Comment (amstewart) [inline src/nilrt-snac:3]: This wrapper script only works for installations that are using `/usr/local` as their prefix. There are a few was to fix this.

1) You (the project designer) can punt on making a wrapper script entirely and leave it up to the distribution vendor (in this case, meta-nilrt). The vendor has knowledge of where the package is being installed and can write their own wrapper.

This effectively punts the burden to distro maintainers. But it means that you won't necessarily have a nice wrapper script for devs who are building locally.

2) You can use a templating variable in place of the prefix like

```bash
export PYTHONPATH=%prefix%/lib/nilrt-snac
python3 -m nilrt_snac $@
```

And then use `sed` or another templating engine during `make` to write the Make `$(prefix)` value into this file.

This gives you a valid wrapper script for local builds of the file, but requires more logic in the Makefile. For larger projects that use Automake and already have workflows to `./configure` and templatize content, it isn't a big deal. But it would mean new steps for ours.

3) You can assert that the wrapper script should always be installed to the same `prefix` as the other project content, and so the prefix can be assumed within a relative path.

```bash
export PYTHONPATH=$(realpath ../lib/nilrt-snac)
python3 -m nilrt_snac $@
```

This should always work, assuming that downstream installers are using our installation logic. But you might argue that if they are customizing the installation logic, it's OK to make them write their own wrapper.

Comment (amstewart) [inline src/nilrt-snac:3]: It's generally preferable to pass arguments using the `$@` variable. Both `$*` and `$@` contain the current shell context arguments, but the former stores them as a space-delimited string and the latter as an array. So the latter preserves tokenization based on quotes. eg.
```
$ script.sh "foo bar" one two
$* = foo, bar, one, two
$@ = foo bar, one, two
```

**Review by amstewart — CHANGES_REQUESTED** [review]

File: nilrt_snac/_Configs/_CryptsetupConfig.py
Comment (amstewart) [reply nilrt_snac/_Configs/_CryptsetupConfig.py]: These are still using lower-case letters in Patchrev 2.

File: nilrt_snac/__main__.py
Comment (texasaggie97) [reply nilrt_snac/__main__.py:86]: It is no longer relevant. Removed.

**Comment by amstewart** [comment]
@texasaggie97 I just pulled #3. So you may need to rebase this PR on the latest `master`.

File: nilrt_snac/requirements.txt
Comment (amstewart) [inline nilrt_snac/requirements.txt:2]: You've dropped the linting target entirely; right? Do you still need these modules?

**Review by amstewart — CHANGES_REQUESTED** [review]

File: src/nilrt-snac
Comment (texasaggie97) [reply src/nilrt-snac]: Changed.

File: nilrt_snac/__init__.py
Comment (texasaggie97) [reply nilrt_snac/__init__.py]: Changed.

File: src/nilrt-snac
Comment (texasaggie97) [reply src/nilrt-snac]: Changed to use realpath, but it is more complicated that what is shown here since realpath only works with cwd.

File: nilrt_snac/__init__.py
Comment (texasaggie97) [reply nilrt_snac/__init__.py:7]: Changed.

File: nilrt_snac/__main__.py
Comment (texasaggie97) [reply nilrt_snac/__main__.py:79]: In order to implement `--version` below in a sane way, I removed the direct argparse requirement that one or the other be set, but then explicitly check for one of them after printing the version and exiting if that was selected.

File: nilrt_snac/__main__.py
Comment (texasaggie97) [reply nilrt_snac/__main__.py:66]: Done.

**Comment by texasaggie97** [comment]
# Patch v3

* Include faillock config (converted from bash)
* Addition review comments
  * Add `--version` parameter
  * Remove requirements.txt file
  * Change `Errors` to inherit from `IntEnum` instead of `Enum`
* Change `verify()` to raise an exception when it finds an issue

**Comment by amstewart** [comment]
@texasaggie97 Are you going to update the _Config class names?

**Review by amstewart — APPROVED** [review]

**Pull request merged**
